Source/Mothership/Mothership.cpp

```diff
@@ -0,0 +1,135 @@
+#include "Mothership.h"
+
+Mothership::Mothership(int argc, char** argv):
+    CommonBase(argc, argv, std::string(csMOTHERSHIPproc),
+               std::string(__FILE__)),
+    backend(PNULL),
+    threading(ThreadComms(this))
+{
+    try
+    {
+        instrumentation = InstrumentationWriter();
+    }
+    catch (InstrumentationException &e)
+    {
+        Post(509, e.message);
+    }
+}
+
+Mothership::~Mothership()
+{
+    /* Tear down MPI function map. */
+    WALKVECTOR(FnMap_t*, FnMapx, F) delete *F;
+
+    /* Tear down backend. */
+    if (backend != PNULL)
+    {
+        debug_print("[MOTHERSHIP] Closing down the compute backend...\n");
+        delete backend;
+        debug_print("[MOTHERSHIP] Backend closed.\n");
+        debug_print("[MOTHERSHIP] Disconnecting from MPI (goodbye world).\n");
+    }
+}
+
+/* Dumps dumpable datastructures to a stream. Note that the CommonBase data
+ * structure is not dumped by this, because it requires a file pointer (and
+ * there's no portable way to interface the two). See the dump handler. */
+void Mothership::dump(std::ofstream* stream)
+{
+    *stream << "Mothership (" << POETS::get_hostname() << ") dump:\n";
+    appdb.dump(stream);
+    superdb.dump(stream);
+}
+
+/* Loads the backend and sets the threads spinning, waiting for them to
+ * exit. */
+void Mothership::go()
+{
+    setup_mpi_hooks();
+    load_backend();
+    threading.go();
+}
+
+/* Sends a message using MPI to the process defined in the message.
+ *
+ * You might reasonably thing this is largely superfluous, but this is here to
+ * aid debugging as an elegant hook for all outgoing MPI communications. The
+ * compiler (assuming it's not bad) will inline it under optimisation.
+ *
+ * The argument is an address to avoid difficulties working with the copy
+ * constructor.  */
+void Mothership::queue_mpi_message(PMsg_p* message)
+{
+    message->Send();
+}
+
+/* Loads the compute backend, crashing violently if it does (for now). */
```

Comment:
I certainly does crash violently, but not by design ;). It won't forever, pending the `HostLink` changes that I said I would make (but haven't done).

Reply:
Am I being dense then? "crashing violently" sounds like a catastrophic failure, which would mean that the compute backend is never loaded?

Is this a terminology issue or is there a "not" missing from the comment?

Reply:
It is a catastrophic failure - HostLink calls the C builtin `exit` when something goes wrong. This ends the Mothership process. See PIP-0025 (at https://github.com/POETSII/poets_improvement_proposals/blob/master/proposed/PIP-0025-hostlink-software-interfaces.md).

To clarify when I write "Loads the compute backend, crashing violently if it does (for now).", "it" is the compute backend. Consider reading the comment like this:

> [this method] Loads the compute backend, crashing violently if the compute backend does (for now).

Reply:
Ah, I was reading the method as being the (implicit) subject rather than the compute backend. e.g. I was reading it as "(This method) loads the compute backend, crashing violently if it does (load)" 

This comment could certainly be clearer, how about:
`/* Loads the compute backend. N.B. if the compute backend crashes, it does so violently.*/`

Reply:
Changed as of `47134aa`.